amandaguan-ag/todo
# Authentication

## Type of Changes



|     | Type                       |
| --- | -------------------------- |
|  ✓  | :sparkles: New feature     |
|     | :bug: Bug fix              |
|     | :hammer: Refactoring       |
|  ✓  | :link: Update dependencies |
|     | :100: Add tests            |

## Acceptance Criteria

- [x] Implement the sign-up and log-in functionalities.
- [x] Setup API endpoints for sign-in and log-in.
- [x] User can register a new account and log in with existing credentials. 

## Related Issue

Closes #14 

## Description

This PR introduces the sign-up and log-in features to enhance user authentication. It also sets up the necessary API endpoints to support these functionalities.

### Changes Made:
1. **Updated Routing Configuration:**
   - Integrated `BrowserRouter` and `createBrowserRouter` from `react-router-dom` to manage our application routes.
   - Configured routes for Home, Sign-Up, and Log-In components.
   - The application now dynamically handles routes for better user navigation and state management.

2. **API Endpoints:**
   - Developed API endpoints for sign-in (`POST /api/login`) and sign-up (`POST /api/signup`).
   - Utilized `bcrypt` for password hashing and `jsonwebtoken` for generating tokens.

## Testing Steps / QA Criteria
To prepare for testing, please ensure you've followed the [setup steps](https://github.com/amandaguan-ag/todo#setup) outlined in the project's README, including running any necessary database migrations to update your local database schema.

1. Checkout to this branch with `git checkout sign-in-api`.
2. Make sure you are in the backend directory and run the database migrations to update your schema by executing `npm run typeorm migration:run`.
3. Start the backend service by running `npm run start` in the backend directory.
4. In a separate terminal, navigate to the frontend directory and run `npm start` to launch the frontend.
5. Navigate to [http://localhost:3000](http://localhost:3000) to access the Todo List Application.
6. Try signing up for a new account and then log in with the same credentials.

## TODOs

- [ ] Refactor and DRY code
- [ ] Add Navbar for better naviagtion
